src/rust/graph-merger/src/main.rs

```diff
@@ -31,64 +47,151 @@ pub mod service;
+    );
+
+    // TODO: also construct a stream processor for retries
+
+    let stream_processor: StreamProcessor<Envelope<IdentifiedGraph>, Envelope<MergedGraph>> =
+        StreamProcessor::new(consumer_config, producer_config)?;
+
+    tracing::info!(message = "kafka stream processor configured successfully",);
+
+    let stream = stream_processor.stream::<_, _, StreamProcessorError>(
+        move |event: Result<Envelope<IdentifiedGraph>, StreamProcessorError>| {
+            let graph_merger = graph_merger.clone();
+            async move {
+                let envelope = event?;
+                match graph_merger
+                    .lock()
+                    .await
+                    .handle_event(envelope.inner_message)
+                    .await
+                {
+                    Ok(merged_graph) => Ok(Some(Envelope::new(
+                        Metadata::create_from(envelope.metadata),
+                        merged_graph,
+                    ))),
+                    Err(e) => match e {
+                        Ok((_, e)) => {
+                            match e {
+                                GraphMergerError::Unexpected(ref reason) => {
+                                    tracing::error!(
+                                        message = "unexpected error",
+                                        reason = %reason,
+                                        error = %e,
+                                    );
+                                    // TODO: write message to failed topic here
+                                    Err(StreamProcessorError::from(e))
+                                }
+                                _ => {
+                                    tracing::error!(
+                                        mesage = "unexpected error",
+                                        error = %e,
+                                    );
+                                    // TODO: write message to failed topic here
+                                    Err(StreamProcessorError::from(e))
+                                }
+                            }
+                        }
+                        Err(e) => match e {
+                            GraphMergerError::Unexpected(reason) => {
+                                tracing::warn!(
+                                    message = "unexpected error",
+                                    reason = %reason,
+                                );
+                                Ok(None)
+                            }
+                            _ => {
+                                tracing::error!(
+                                    message = "unknown error",
+                                    error = %e,
+                                );
+                                Err(StreamProcessorError::from(e))
+                            }
+                        },
+                    },
+                }
+            }
+        },
+    )?;
```

Comment:
I won't push the matter, but I'd argue this is a pretty code-smell-y level of indentation